examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/ntp.c

```diff
@@ -0,0 +1,301 @@
+   *         API callbacks. It will be set to the socket ID returned by connect()
+   *         - epoch_time will be set to the current epoch time as returned by the
+   *         SNTP server
+   * @retval sntp_status_t: 0 means SNTP_SUCCESS, <0 means something failed.
+   *         See ntp.h to handle different failure reasons in different ways
+   */
+sntp_status_t sntp_get_datetime( uint8_t* sock_id, int32_t* epoch_time )
+{
+    sntp_status_t retval = SNTP_SUCCESS; //Returned by this function
+    WiFi_Status_t wifi_retval = WiFi_MODULE_SUCCESS;
+
+    /* Create socket */
+    printf("\r\n>>Getting date and time from SNTP server...");
+    wifi_retval = sntp_connect(SNTP_SERVER, SNTP_PORT, sock_id);
+    if( wifi_retval != WiFi_MODULE_SUCCESS )
+    {
+        printf("\r\n\tSNTP socket connection [FAIL] Retval: %d", wifi_retval);
+        retval = SNTP_CONNECTION_FAILURE;
+        goto terminate;
+    }
+    else if(*sock_id < 0)
+    {
+        printf("\r\n\tUDP Connection [FAIL] Socket ID<0: %d", *sock_id);
+        retval = SNTP_CONNECTION_FAILURE;
+        goto terminate;
+    }
+    printf("\r\n\tUDP Connection [OK] Assigned socket ID: %d", *sock_id);
+
+    /* Send SNTP request */
+    wifi_retval = sntp_send_request(*sock_id);
+    if( wifi_retval != WiFi_MODULE_SUCCESS )
+    {
+        printf("\r\n\tSNTP send_request [FAIL] Retval: %d", wifi_retval);
+        retval = SNTP_REQUEST_FAILURE;
+        goto terminate;
+    }
+
+    /* Await SNTP Response */
+    printf("\r\n\tAwaiting server response");
+    if( sntp_await_response(*sock_id) < 0 )
+    {
+        printf("\r\n\tSNTP response [FAIL] Response timed out");
+        retval = SNTP_TIMEOUT;
+        goto terminate;
+    }
+
+    if( *sock_id != last_sntp_response->socket_id )
+    {
+        printf("\r\n\tSocket ID assertion [FAIL] Bug in ntp.c?");
+        retval = SNTP_INTERNAL_ERROR;
+        goto terminate;
+    }
+
+    /* Parse server response */
+    printf("\r\n\tParsing and converting SNTP response");
+    sntp_current_time = sntp_parse_response(last_sntp_response->response);
+    if( sntp_current_time < 0 )
+    {
+        printf("\r\n\tSNTP Parsing and conversion [FAIL]");
+        retval = SNTP_PARSER_ERROR;
+    }
+    else
+    {
+        printf("\r\n\tReceived epoch time: %ld", sntp_current_time);
+        retval = SNTP_SUCCESS;
+    }
+
+terminate:
+    /* free() SNTP Response */
+    if( NULL != last_sntp_response )
+    {
+        sntp_free_response(last_sntp_response);
+        last_sntp_response = NULL;
+    }
+
+    /* Close socket */
+    if( *sock_id > 0 )
+    {
+        sntp_disconnect(*sock_id);
+    }
+    return retval;
+}
+
+/**
+   * @brief
+   * @param  sock_id we're interested in
+   * @retval SNTP_SUCCESS (0) or SNTP_TIMEOUT(-1)
+   */
+sntp_status_t sntp_await_response( uint8_t sock_id )
+{
+    int32_t sntp_timeout_ms = SNTP_TIMEOUT_MS;
+    const int32_t timeout_step = 250;
+    while( NULL == last_sntp_response )
```

Comment:
There is a probability that this implementation will leak the memory - if the response from the server will come after SNTP_TIMEOUT occurs the ```sntp_socket_data_callback``` will allocate the memory and this memory will not be released - The ```sntp_socket_data_callback``` function must also check if the request is still pending, if not it suppose to send the data to the /dev/null.